client/wfm/src/service/WfmService.ts

```diff
@@ -0,0 +1,248 @@
+import * as Promise from 'bluebird';
+import * as _ from 'lodash';
+import * as shortid from 'shortid';
+import { WorkOrderResult } from '../result/WorkOrderResult';
+import { STATUS } from '../status';
+import { Step } from '../step/Step';
+import { WorkFlow } from '../workflow/WorkFlow';
+import { WorkOrder } from '../workorder/WorkOrder';
+import { DataService } from './DataService';
+import { ResultService } from './ResultService';
+import { UserService } from './UserService';
+
+/**
+ * Parameter interface for the {@link WfmService#completeStep()} function
+ */
+export interface CompleteStepParams {
+  /** The ID of the workorder to complete the step for */
+  workorderId: string;
+  /** The submission to save */
+  submission: object;
+  /** The ID of the step to save the submission for */
+  stepCode: string;
+}
+
+/**
+ * The return value of one of the operations that change the active {@link Step} of a {@link WorkOrder}
+ */
+export interface StepOperationData extends Summary {
+  step: Step;
+}
+
+/**
+ * Return value of the {@link WfmService#workorderSummary} interface
+ */
+export interface Summary {
+  workflow: WorkFlow;
+  workorder: WorkOrder;
+  result: WorkOrderResult;
+  /**
+   * Index of the next {@link Step} the {@link WorkOrder} should progress to after completion of the current one.
+   * This allows for skipping or returning to a previous one.
+   */
+  nextStepIndex: number;
+}
+
+export class WfmService {
+
+  constructor(
+    protected workorderService: DataService<WorkOrder>,
+    protected workflowService: DataService<WorkFlow>,
+    protected resultService: ResultService,
+    protected userService: UserService
+  ) {
+  }
+
+  /**
+   * Beginning a workflow for a single Workorder.
+   *
+   * @param {string} workorderId - The ID of the workorder to begin the workflow for.
+   */
+  public beginWorkflow(workorderId): Promise<StepOperationData> {
+    return this.workorderSummary(workorderId).then(summary => {
+      if (summary.result) {
+        console.warn(`beginWorkflow() called on already started workflow with id: ${workorderId},`
+          + ` ignoring existing result`);
+      }
+      const { workorder, workflow, nextStepIndex } = summary;
+      return this.createNewResult(workorderId, workorder.assignee)
+        .then(result => {
+          // Now we check the current status of the workflow to see where the next step should be.
+          result.status = this.checkStatus(workorder, workflow, result);
+
+          // We now have the current status of the workflow for this workorder, the begin step is now complete.
+          return {
+            workorder,
+            workflow,
+            result,
+            nextStepIndex,
+            step: nextStepIndex > -1 ? workflow.steps[nextStepIndex] : workflow.steps[0]
```

Comment:
Extract this to method or improve - it's really hard to read and extract what is happening here.